### PR TITLE
fix: broken integ tests and UPDATED_NEW/OLD multiple subpaths

### LINKED ISSUE #98: [Bug]  update_item `UPDATED_NEW` and `UPDATED_OLD` return too much data
## Describe the bug

ReturnValues UPDATED_NEW and UPDATED_OLD  return more data than in DynamoDB when mutating nested data. DynamoDB returns only the changed fragment while ExtendDB is returning entire parent maps or lists. For non-nested lists ExtendDB is returning the whole list when a SET operation is used to insert new data, compared to DynamoDB returning no data

## To Reproduce

**I had an LLM generate this text from unit test failures to save my fingers**

For each scenario, insert this item

```json
{
  "TableName": "test-table",
  "Item": {
    "pk": {
      "S": "a"
    },
    "m": {
      "M": {
        "a": {
          "M": {
            "b": {
              "S": "old-b"
            },
            "c": {
              "N": "1"
            }
          }
        },
        "keep": {
          "S": "mk"
        }
      }
    },
    "l": {
      "L": [
        {
          "M": {
            "a": {
              "S": "old0"
            },
            "b": {
              "S": "b0"
            }
          }
        },
        {
          "M": {
            "a": {
              "S": "old1"
            },
            "b": {
              "S": "b1"
            }
          }
        }
      ]
    },
    "rootList": {
      "L": [
        {
          "S": "x"
        },
        {
          "S": "y"
        }
      ]
    }
  }
}
```

### Execute Request

```json
{
  "TableName": "test-table",
  "Key": {
    "pk": {
      "S": "a"
    }
  },
  "UpdateExpression": "SET m.a.b = :v",
  "ExpressionAttributeValues": {
    ":v": {
      "S": "new-b"
    }
  },
  "ReturnValues": "UPDATED_NEW"
}
```

### Expected DynamoDB Response

```json
{
  "Attributes": {
    "m": {
      "M": {
        "a": {
          "M": {
            "b": {
              "S": "new-b"
            }
          }
        }
      }
    }
  }
}
```

### Actual ExtendDB Response

```json
{
  "Attributes": {
    "m": {
      "M": {
        "a": {
          "M": {
            "b": {
              "S": "new-b"
            },
            "c": {
              "N": "1"
            }
          }
        },
        "keep": {
          "S": "mk"
        }
      }
    }
  }
}
```

### Execute Request

```json
{
  "TableName": "test-table",
  "Key": {
    "pk": {
      "S": "a"
    }
  },
  "UpdateExpression": "SET m.a.b = :v",
  "ExpressionAttributeValues": {
    ":v": {
      "S": "new-b"
    }
  },
  "ReturnValues": "UPDATED_OLD"
}
```

### Expected DynamoDB Response

```json
{
  "Attributes": {
    "m": {
      "M": {
        "a": {
          "M": {
            "b": {
              "S": "old-b"
            }
          }
        }
      }
    }
  }
}
```

### Actual ExtendDB Response

```json
{
  "Attributes": {
    "m": {
      "M": {
        "a": {
          "M": {
            "b": {
              "S": "old-b"
            },
            "c": {
              "N": "1"
            }
          }
        },
        "keep": {
          "S": "mk"
        }
      }
    }
  }
}
```

### Execute Request

```json
{
  "TableName": "test-table",
  "Key": {
    "pk": {
      "S": "a"
    }
  },
  "UpdateExpression": "SET l[1].a = :v",
  "ExpressionAttributeValues": {
    ":v": {
      "S": "new1"
    }
  },
  "ReturnValues": "UPDATED_NEW"
}
```

### Expected DynamoDB Response

```json
{
  "Attributes": {
    "l": {
      "L": [
        {
          "M": {
            "a": {
              "S": "new1"
            }
          }
        }
      ]
    }
  }
}
```

### Actual ExtendDB Response

```json
{
  "Attributes": {
    "l": {
      "L": [
        {
          "M": {
            "a": {
              "S": "old0"
            },
            "b": {
              "S": "b0"
            }
          }
        },
        {
          "M": {
            "a": {
              "S": "new1"
            },
            "b": {
              "S": "b1"
            }
          }
        }
      ]
    }
  }
}
```

### Execute Request

```json
{
  "TableName": "test-table",
  "Key": {
    "pk": {
      "S": "a"
    }
  },
  "UpdateExpression": "SET l[1].a = :v",
  "ExpressionAttributeValues": {
    ":v": {
      "S": "new1"
    }
  },
  "ReturnValues": "UPDATED_OLD"
}
```

### Expected DynamoDB Response

```json
{
  "Attributes": {
    "l": {
      "L": [
        {
          "M": {
            "a": {
              "S": "old1"
            }
          }
        }
      ]
    }
  }
}
```

### Actual ExtendDB Response

```json
{
  "Attributes": {
    "l": {
      "L": [
        {
          "M": {
            "a": {
              "S": "old0"
            },
            "b": {
              "S": "b0"
            }
          }
        },
        {
          "M": {
            "a": {
              "S": "old1"
            },
            "b": {
              "S": "b1"
            }
          }
        }
      ]
    }
  }
}
```

### Execute Request

```json
{
  "TableName": "test-table",
  "Key": {
    "pk": {
      "S": "a"
    }
  },
  "UpdateExpression": "REMOVE m.a.b",
  "ReturnValues": "UPDATED_NEW"
}
```

### Expected DynamoDB Response

```json
{}
```

### Actual ExtendDB Response

```json
{
  "Attributes": {
    "m": {
      "M": {
        "a": {
          "M": {
            "c": {
              "N": "1"
            }
          }
        },
        "keep": {
          "S": "mk"
        }
      }
    }
  }
}
```

### Execute Request

```json
{
  "TableName": "test-table",
  "Key": {
    "pk": {
      "S": "a"
    }
  },
  "UpdateExpression": "REMOVE m.a.b",
  "ReturnValues": "UPDATED_OLD"
}
```

### Expected DynamoDB Response

```json
{
  "Attributes": {
    "m": {
      "M": {
        "a": {
          "M": {
            "b": {
              "S": "old-b"
            }
          }
        }
      }
    }
  }
}
```

### Actual ExtendDB Response

```json
{
  "Attributes": {
    "m": {
      "M": {
        "a": {
          "M": {
            "b": {
              "S": "old-b"
            },
            "c": {
              "N": "1"
            }
          }
        },
        "keep": {
          "S": "mk"
        }
      }
    }
  }
}
```

### Execute Request

```json
{
  "TableName": "test-table",
  "Key": {
    "pk": {
      "S": "a"
    }
  },
  "UpdateExpression": "SET rootList[99] = :v",
  "ExpressionAttributeValues": {
    ":v": {
      "S": "z"
    }
  },
  "ReturnValues": "UPDATED_NEW"
}
```

### Expected DynamoDB Response

```json
{}
```

### Actual ExtendDB Response

```json
{
  "Attributes": {
    "rootList": {
      "L": [
        {
          "S": "x"
        },
        {
          "S": "y"
        },
        {
          "S": "z"
        }
      ]
    }
  }
}
```

### Execute Request

```json
{
  "TableName": "test-table",
  "Key": {
    "pk": {
      "S": "sa"
    }
  },
  "UpdateExpression": "SET rootList[99] = :v",
  "ExpressionAttributeValues": {
    ":v": {
      "S": "z"
    }
  },
  "ReturnValues": "UPDATED_OLD"
}
```

### Expected DynamoDB Response

```json
{}
```

### Actual ExtendDB Response

```json
{
  "Attributes": {
    "rootList": {
      "L": [
        {
          "S": "x"
        },
        {
          "S": "y"
        }
      ]
    }
  }
}
```

## Environment

- ExtendDB version: 0.1.0
- Operating system: macOS
- Rust version (if building from source): 1.94
- Client SDK/driver and version: CLI
- Deployment method (binary, container, source): source

## Checklist

- [x] I have searched existing issues for duplicates
- [x] I have included the ExtendDB version and environment details
- [x] I can reproduce this with the latest release

### PR DESCRIPTION
**Pull request by LeeroyHannigan**
## What
  

1. **Fix 11 broken integration tests** that were failing on both real DynamoDB and ExtendDB due to test bugs (reserved keywords used without `#name` placeholders, wrong error code assertions, wrong account in ARN).

2. **Fix UPDATED_NEW/OLD with multiple subpaths** under the same top-level attribute (e.g., `SET a.b = :x, a.c = :y` now returns both `a.b` and `a.c` instead of only the last one).

3. **Fix 4 additional engine bugs** found during investigation:
   - TransactWriteItems now charges 2x WCU (transaction cost), matching DynamoDB
   - Scan with negative Segment returns the correct DynamoDB-format ValidationException
   - UpdateTable deleting a non-existent GSI returns ResourceNotFoundException with correct message
   - `contains(#a, #a)` error message now resolves `#name` placeholders to the actual attribute name

4. **Add 4 new integration tests** for the engine fixes above so they don't regress.

## Why

Closes #98 (the remaining edge case from #105)

The 11 test failures were masking real progress, they failed on real DynamoDB too, meaning the tests themselves were wrong, not the engine. Fixing them gives us a clean signal: 771/776 pytest passing.

## Testing done

- All 11 fixed tests verified against **real DynamoDB** first (all pass)
- Then verified against ExtendDB (all pass)
- 4 new tests verified against ExtendDB (all pass)
- Error messages for scan segment, UpdateTable GSI, and contains() verified against real DynamoDB for exact match
- `cargo test --workspace` — all pass
- `cargo fmt --all -- --check` — clean
- `cargo clippy --all-targets -- -D warnings` — clean
- Full pytest: 771 passed, 5 failed (import/export disabled in config only)

## Checklist

- [x] I have read [CONTRIBUTING.md](../CONTRIBUTING.md)
- [x] All tests pass (`cargo test --workspace`)
- [x] Code is formatted (`cargo fmt --check`)
- [x] Clippy is clean (`cargo clippy -- -W clippy::pedantic`)
- [x] I have added or updated tests for new functionality
- [ ] I have updated documentation if behavior changed
- [ ] Breaking changes are noted below (if any)
- [ ] If this changes the wire protocol, `Storage` trait, auth model, on-disk
      format, or public CLI surface, an RFC has been accepted or is linked
      below. Otherwise, an ADR captures the decision (link below).

ADR / RFC: n/a

---

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache License 2.0 and I agree to the Developer Certificate of
Origin (DCO). See [CONTRIBUTING.md](../CONTRIBUTING.md) for details.

### CODE REVIEW

File: crates/engine/src/update_item.rs [hunk 1]
```diff
@@ -325,7 +325,17 @@ fn filter_to_updated_attrs(item: &Item, actions: &[UpdateAction], maps: &Express
         };
         if let Some(leaf) = resolve_path_value(top_val, &path[1..], maps) {
             let wrapped = wrap_leaf_in_path(&path[1..], &leaf, maps);
-            result.insert(top_name, wrapped);
+            // Merge into existing top-level entry if multiple subpaths target
+            // the same attribute (e.g., SET a.b = :x, a.c = :y).
+            if let Some(existing) = result.get(&top_name) {
+                if let Some(merged) = merge_maps(existing, &wrapped) {
+                    result.insert(top_name, merged);
```
Comment (pdf-amzn) [inline crates/engine/src/update_item.rs:332]: seems like this is just `result.insert(top_name, merge_maps(existing, &wrapped).or_else(wrapped))` or something like that.

File: crates/engine/src/update_item.rs [hunk 2]
```diff
@@ -380,6 +390,28 @@ fn wrap_leaf_in_path(
     }
 }
 
+/// Recursively merge two Map AttributeValues. Returns None if either isn't a Map.
+fn merge_maps(a: &AttributeValue, b: &AttributeValue) -> Option<AttributeValue> {
```
Comment (pdf-amzn) [inline crates/engine/src/update_item.rs:394]: In isolation, it seems like this should return `Error` instead of `Option`, even though `Option` is a good fit for where you actually use this. Interesting tradeoff. Definitely nonblocking though!

File: crates/engine/src/update_table.rs [hunk 3]
```diff
@@ -164,7 +164,8 @@ pub async fn handle_update_table(
             }
             extenddb_storage::error::StorageError::IndexNotFound(name) => {
                 DynamoDbError::ResourceNotFoundException(format!(
-                    "One or more parameter values were invalid: Index not found: {name}"
+                    "Requested resource not found: Index {name} for table {}",
```
Comment (pdf-amzn) [inline crates/engine/src/update_table.rs:167]: I'm surprised clippy didn't ask you to use `{table_name}`

**Review by pdf-amzn — APPROVED** [review]
a few comments, nothing blocking.
